### ISSUE TITLE
Allow `allowImportingTsExtensions`

### ISSUE THREAD

**Issue by Zamiell** [TypeScriptToLua/TypeScriptToLua#1514]
It makes sense in TSTL projects to import from .ts file extensions, like so:

```ts
import { foo } from "./foo.ts";
```

However, when doing this, TSTL will give the following error:

```
src/main.ts(1,21): error TS5097: An import path can only end with a '.ts' extension when 'allowImportingTsExtensions' is enabled.
```

Note that this error is spurious; TSTL will actually still go on to compile correct Lua output (and declaration maps), from what I can tell.

In order to get rid of the spurious error, we can turn on 'allowImportingTsExtensions', but then we get another error:

```
tsconfig.json(25,39): error TS5096: Option 'allowImportingTsExtensions' can only be used when either 'noEmit' or 'emitDeclarationOnly' is set.
```

Similar to the first error, this error is also spurious, meaning that Lua files will properly be created. However, if we enable either the `noEmit` or `emitDeclarationOnly` compiler flags, now TSTL will actively refuse to generate any Lua files.

Thus, we are stuck, and we need some change in TSTL itself.

Proposition: I think that TSTL should customize the behavior of TypeScript in this case and allow `allowImportingTsExtensions` without enabling `noEmit` or `emitDeclarationOnly`.

Another potential solution is to have `allowImportingTsExtensions` turned on by default for all TSTL projects, to make it easier for end-users.

**Comment by Z3rio** [comment]
Isn't this warning related to the actual typescript compiler, not TSTL?
Imports containing `.ts` shouldn't be used in TSTL nor normal TS afaik.

**Comment by Perryvw** [comment]
What's the added benefit of allowing this?

**Comment by Zamiell** [comment]
> Isn't this warning related to the actual typescript compiler, not TSTL?

Sort of. The message comes from the TypeScript compiler because Node.js / a browser can't resolve TypeScript imports at run-time, and since the TypeScript compiler itself won't arbitrarily change file extensions from ".ts" to ".js", then those extensions would always result in broken code. However, the ecosystem doesn't use TypeScript to compile code anymore, only to type check (since it is so slow). The actual bundlers (e.g. esbuild, swc, vite, etc.) can make ".ts" file extensions work just fine, which is why TypeScript expects you to turn "noEmit" on and use it purely for type-checking, and then use the bundler of your actual choice in order to actually transpile/bundle the code.

> Imports containing .ts shouldn't be used in TSTL nor normal TS afaik.

Incorrect. New TypeScript projects should be using [ESM](https://nodejs.org/api/esm.html), and in ESM, you should use file extensions in import statements. In fact, [the recommended TSConfig turns on `isolatedModules`](https://github.com/tsconfig/bases/blob/main/bases/strictest.json), which enforces this at the TypeScript compiler level. And for projects that are completely written in TypeScript, it doesn't make sense to use ".js" file extensions, because the resulting JavaScript files might not ever exist, so it also makes sense to use ".ts" file extensions and turn on the "allowImportingTsExtensions" compiler flag.

> What's the added benefit of allowing this?

As you are probably aware, the TypeScript ecosystem is transitioning to ESM. If a server is written in TypeScript, the benefit of writing a Lua client in TypeScript is that both server and client can share code - this is probably one of the core use-cases for TSTL. However, modern servers are going to be written in ESM, and can only use ESM deps, and can't easily consume CJS code. So people transitioning their TSTL code from CJS to ESM are going to run into this issue.

Even if our TSTL code is never consumed by the outside world, there is arguably still a benefit to having our TSTL codebases work with modern, idiomatic TypeScript code, in that there will be less "gotchas" for onboarding TSTL users who are used to writing TypeScript the proper way.

**Comment by Perryvw** [comment]
Looking into it a bit more I think this is a larger issue I'd rather not burn my fingers on. Some disambiguation:

* `allowImportingTsExtensions` is NOT ECMAScript modules. TypeScript already supports Ecmascript modules, done by having `"type": "module"` in package.json and `"module": "nodenext"` in tsconfig. (See [typescript docs](https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-nodenext))
* ECMAScript modules require imports to have `.js` extensions, not `.ts`
* ECMAScript modules are a _compilation target_, not a decriptor of the _source code_.
  * Since TSTL has a different _target_ (Lua), this seems out of scope for this project. 

As for your original question:
* The error you are giving is a TypeScript diagnostic on your tsconfig not being valid. I don't quite understand why this exists but it's coming from TS and has nothing to do with TSTL. We could maybe allow this, but it does not actually bring what you are claiming it to, because your code would not be valid ECMAScript modules, it would just be imports with .ts extensions (while ESM needs .js).

**Comment by Zamiell** [comment]
> We could maybe allow this

Can you allow it please? I don't think it is right to force TSTL users to use ".js" file extensions. That would definitely seem like a bug in TSTL, because obviously there's never any JavaScript involved, its just directly TypeScript to Lua.

> but it does not actually bring what you are claiming it to

This statement confuses me. As I said in the OP, it does bring what I am claiming it to, because TSTL seems to produce correct output when I have ".ts" file extensions in my TSTL project. All we need is the spurious error to be silenced by TSTL. Or is the output somehow bugged too?

> because your code would not be valid ECMAScript modules, it would just be imports with .ts extensions (while ESM needs .js)

From what I understand, this statement is false. ESM does not require .js, it only requires a file extension. For example, it is valid ESM to import a JSON file, and there is even [a TC39 proposal to use import type assertions](https://github.com/tc39/proposal-json-modules).
